File: class6/ping_pong.py
from mpi4py import MPI
comm = MPI.COMM_WORLD
rank = comm.Get_rank()
MAX = 1000
i = 0
while i < MAX:
    if rank == 0:
        print(f"ping, {i}")
        i+=1
        comm.send(i, dest = 1, tag=0)
        i = comm.recv(source = 1, tag=0)
    else:
        i = comm.recv(source = 0, tag=0)
        print(f"pong, {i}")
        i+=1
        comm.send(i, dest = 0, tag=0)

File: class6/ping_pong.c
#include <mpi.h>
#include <stdio.h>
#define MAX 1000

int main(int argc, char **argv)
{
  MPI_Init(&argc, &argv);
  int rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  MPI_Status status;
  int i = 0;
  while(i < MAX)
  {
    if(rank == 0)
    {
      printf("ping\n");
      ++i;
      MPI_Send(&i, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
      MPI_Recv(&i, 1, MPI_INT, 1, 0, MPI_COMM_WORLD, &status);
    }
    else
    {
      MPI_Recv(&i, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, &status);
      printf("pong\n");
      ++i;
      MPI_Send(&i, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
    }
  }
  
  MPI_Finalize();
  return 0;
}
